### PR TITLE
SI-5353 unsound Array lubs.

### PR DESCRIPTION
**Pull request by paulp**
mergePrefixAndArgs tried too hard to finesse the Array lub. When
there is more than one distinct Array element type being lubbed,
abandon the attempt to erase to any common Array type, because the
storage must match all the element types and Nothing will always
break this. Simply erase to AnyRef.

### CODE REVIEW

**Comment by paulp** [comment]
Opened against master because I'm not sure about the implications for compatibility. Inferring different lubs can lead to different method descriptors.

**Comment by paulp** [comment]
Review by @xeno-by, @retronym

**Comment by xeno-by** [comment]
Could you please comment on why the fix should target `lub`, but not `Erasure` (in `unboundedGenericArrayLevel`)? It seems useful to have lubs as precise as possible and downgrade to `Object`s only when necessary. This probably has some implications wrt type inference, which know almost nothing about, so I'm sorry if the question is being stupid.

**Comment by paulp** [comment]
It's possible you're right. I was already writing a mail asking about why Types knows anything about erasure at all. But the primary reason I pursued that location is that mergePrefixAndArgs is already intervening attempting to avoid exactly this problem.

**Comment by paulp** [comment]
But what is the case you have in mind where this commit leads to a less precise lub than is warranted?

**Comment by xeno-by** [comment]
No particular case, just trying to figure out how this should work. I don't know much about advanced type machinery.

**Comment by xeno-by** [comment]
Well, it could matter for macros of all sorts, including type tag materialization, but since the change only affects erasure, it doesn't matter.

**Comment by paulp** [comment]
You're right at least that this is insufficient; I can still induce CCEs.

**Comment by xeno-by** [comment]
Yeah, same here. E.g. see `files/run/t5353.scala`

**Comment by paulp** [comment]
How about now.

**Comment by xeno-by** [comment]
```
object Test extends App {
  def f(x: Boolean) = if (x) Array("abc") else Array()
  println(f(false).length)
}
```

```
uncaught exception during compilation: RuntimeException("scala.Array.length : Object, Object, Any") @ scala.sys.package$.error(package.scala:27)
error: java.lang.RuntimeException: scala.Array.length : Object, Object, Any
    at scala.sys.package$.error(package.scala:27)
    at scala.tools.nsc.backend.ScalaPrimitives$$anonfun$elementType$1$1$$anonfun$apply$1.apply(ScalaPrimitives.scala:571)
    at scala.tools.nsc.backend.ScalaPrimitives$$anonfun$elementType$1$1$$anonfun$apply$1.apply(ScalaPrimitives.scala:571)
    at scala.Option.getOrElse(Option.scala:120)
    at scala.tools.nsc.backend.ScalaPrimitives$$anonfun$elementType$1$1.apply(ScalaPrimitives.scala:571)
    at scala.tools.nsc.backend.ScalaPrimitives$$anonfun$elementType$1$1.apply(ScalaPrimitives.scala:567)
    at scala.reflect.internal.SymbolTable.enteringPhase(SymbolTable.scala:214)
    at scala.tools.nsc.Global.enteringTyper(Global.scala:1056)
    at scala.tools.nsc.backend.ScalaPrimitives.elementType$1(ScalaPrimitives.scala:567)
    at scala.tools.nsc.backend.ScalaPrimitives.getPrimitive(ScalaPrimitives.scala:607)
    at scala.tools.nsc.backend.icode.GenICode$ICodePhase.genPrimitiveOp(GenICode.scala:428)
```

**Comment by xeno-by** [comment]
Looks like the change has to be applied to `unboundedGenericArrayLevel` as that function is used elsewhere.

**Comment by paulp** [comment]
So I am discovering.

**Comment by paulp** [comment]
Now it's definitely overly broad, but at least maybe it will work. I'll refine it.

File: src/reflect/scala/reflect/internal/transform/Erasure.scala [hunk 1]
```diff
@@ -120,12 +123,12 @@ trait Erasure {
         tp
       case st: SubType =>
         apply(st.supertype)
+      case TypeRef(pre, ArrayClass, arg :: Nil) =>
+        if (unboundedGenericArrayLevel(tp) == 1) ObjectTpe
+        else if (arg.typeSymbol.isBottomClass) ObjectArray
+        else typeRef(apply(pre), ArrayClass, applyInArray(arg) :: Nil)
```
Comment (xeno-by) [inline src/reflect/scala/reflect/internal/transform/Erasure.scala:129]: Looks like this change can be dropped now?

**Comment by paulp** [comment]
Well this is interestingly brutal to get right. Closing for now, will have to revisit.

**Pull request closed without merging**